File: src/engine/clues.ts
import type { Clue } from "./types";

/** Run-length encode the filled runs of a line. `[]` for an all-empty line. */
export function cluesForLine(line: boolean[]): Clue {
  const clue: number[] = [];
  let run = 0;
  for (const filled of line) {
    if (filled) {
      run++;
    } else if (run > 0) {
      clue.push(run);
      run = 0;
    }
  }
  if (run > 0) clue.push(run);
  return clue;
}

/** Derive row and column clues from a solution grid. */
export function cluesForGrid(solution: boolean[][]): {
  rowClues: Clue[];
  colClues: Clue[];
} {
  const height = solution.length;
  const width = height > 0 ? solution[0].length : 0;
  const rowClues = solution.map(cluesForLine);
  const colClues: Clue[] = [];
  for (let c = 0; c < width; c++) {
    const column: boolean[] = [];
    for (let r = 0; r < height; r++) column.push(solution[r][c]);
    colClues.push(cluesForLine(column));
  }
  return { rowClues, colClues };
}

File: apple/ClueweaveKit/Sources/ClueweaveKit/Clues.swift
// Clue derivation (port of src/engine/clues.ts).

/// Run-length encode the filled runs of a line. `[]` for an all-empty line.
public func cluesForLine(_ line: [Bool]) -> Clue {
    var clue: [Int] = []
    var run = 0
    for filled in line {
        if filled {
            run += 1
        } else if run > 0 {
            clue.append(run)
            run = 0
        }
    }
    if run > 0 { clue.append(run) }
    return clue
}

/// Derive row and column clues from a solution grid.
public func cluesForGrid(_ solution: [[Bool]]) -> (rowClues: [Clue], colClues: [Clue]) {
    let height = solution.count
    let width = height > 0 ? solution[0].count : 0
    let rowClues = solution.map(cluesForLine)
    var colClues: [Clue] = []
    colClues.reserveCapacity(width)
    for c in 0..<width {
        var column: [Bool] = []
        column.reserveCapacity(height)
        for r in 0..<height { column.append(solution[r][c]) }
        colClues.append(cluesForLine(column))
    }
    return (rowClues, colClues)
}
